Mathlib/Tactic/MkIffOfInductiveProp.lean

```diff
@@ -0,0 +1,383 @@
+/-
+Copyright (c) 2018 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, David Renshaw
+-/
+import Lean
+import Mathlib.Tactic.Existsi
+import Mathlib.Tactic.LeftRight
+
+/-!
+# mk_iff_of_inductive_prop
+
+This file defines a tactic `tactic.mk_iff_of_inductive_prop` that generates `iff` rules for
+inductive `Prop`s. For example, when applied to `List.Chain`, it creates a declaration with
+the following type:
+
+```lean
+∀{α : Type _} (R : α → α → Prop) (a : α) (l : List α),
+  Chain R a l ↔ l = [] ∨ ∃(b : α) (l' : list α), R a b ∧ Chain R b l ∧ l = b :: l'
+```
+
+This tactic can be called using either the `mk_iff_of_inductive_prop` user command or
+the `mk_iff` attribute.
+-/
+
+namespace Mathlib.Tactic.MkIff
+
+open Lean Meta
+
+/-- `select m n` runs `tactic.right` `m` times, and then `tactic.left` `(n-m)` times.
+Fails if `n < m`. -/
+private def select (m n : Nat) (goal : MVarId) : MetaM MVarId :=
+  match m,n with
+  | 0, 0             => pure goal
+  | 0, (_ + 1)       => do let [new_goal] ← Mathlib.Tactic.LeftRight.leftRightMeta `left 0 2 goal
+                                | throwError "expected only one new goal"
+                           pure new_goal
+  | (m + 1), (n + 1) => do let [new_goal] ← Mathlib.Tactic.LeftRight.leftRightMeta `right 1 2 goal
+                                | throwError "expected only one new goal"
+                           select m n new_goal
+  | _, _             => failure
+
+/-- `compactRelation bs as_ps`: Produce a relation of the form:
+```lean
+R := λ as ∃ bs, Λ_i a_i = p_i[bs]
+```
+This relation is user-visible, so we compact it by removing each `b_j` where a `p_i = b_j`, and
+hence `a_i = b_j`. We need to take care when there are `p_i` and `p_j` with `p_i = p_j = b_k`.
+-/
+partial def compactRelation :
+  List Expr → List (Expr × Expr) → List (Option Expr) × List (Expr × Expr) × (Expr → Expr)
+| [],    as_ps => ([], as_ps, id)
+| b::bs, as_ps =>
+  match as_ps.span (λ⟨_,p⟩ => ¬ p == b) with
+    | (_, []) => -- found nothing in ps equal to b
+        let (bs, as_ps', subst) := compactRelation bs as_ps
+        (b::bs, as_ps', subst)
+    | (ps₁, (a, _) :: ps₂) => -- found one that matches b. Remove it.
+      let i := fun e => e.replaceFVar b a
+      let (bs, as_ps', subst) := compactRelation (bs.map i) ((ps₁ ++ ps₂).map (λ⟨a, p⟩ => (a, i p)))
+      (none :: bs, as_ps', i ∘ subst)
+
+/-- Generates an expression of the form `∃(args), inner`. `args` is assumed to be a list of fvars.
+When possible, `p ∧ q` is used instead of `∃(_ : p), q`. -/
+def mkExistsLst (args : List Expr) (inner : Expr) : MetaM Expr :=
```

Comment:
Any reason to spell `List` as `Lst` here? We don't usually do this.

Reply:
I was matching the name of the mathlib3 version: https://github.com/leanprover-community/mathlib/blob/1499a80b158f835d31d22d6a6770476abdc53a60/src/tactic/core.lean#L76

I agree that adding the `i` makes it better, so I made that change just now.